### ISSUE TITLE
wav2letter inference

### ISSUE THREAD

**Issue by Ir0098** [ARM-software/ML-zoo#24]
Hello, 

I am trying to evaluate the quantized wav2letter model provided at:

[project link](https://github.com/ARM-software/ML-zoo/tree/master/models/speech_recognition/wav2letter/tflite_int8)

It also includes an input sample and its corresponding output:

- input: [testing_input/input_2_int8/0.npy](https://github.com/ARM-software/ML-zoo/tree/master/models/speech_recognition/wav2letter/tflite_int8/testing_input/input_2_int8)
- output: [testing_output/Identity_int8/0.npy](https://github.com/ARM-software/ML-zoo/tree/master/models/speech_recognition/wav2letter/tflite_int8/testing_output/Identity_int8)

Also, it is suggested to decode the class probabilities (output) using algorithms like beam search and explicitly it is refereed to
**[ctc_beam_search_decoder](https://www.tensorflow.org/api_docs/python/tf/nn/ctc_beam_search_decoder)**  from tensorflow.  
 
Accordingly, as the [model](https://github.com/ARM-software/ML-zoo/blob/master/models/speech_recognition/wav2letter/tflite_int8/wav2letter_int8.tflite) is already quantized, the output tensor is dequantized using the the **_scales_** and **_zero_points_**.

- scales: 0.00390625
- zero_points: -128

Then, the dequntized output is generated as  below:

```
deq_output = (output.astype(np.float32)-zero_points) * scales

```
And finally, it is passed to the decoder in order to create the sequence of letters:

```
# batch size: 1
# deq_output.shape after reshaping: (148, 1, 29) ==> (sequence length, batch size, # classes/characters)
# sequence_length : 148

(decoded, log_probabilities) = tf.nn.ctc_beam_search_decoder(inputs=deq_output,
                          sequence_length=[sequence_length],
                          beam_width=25,
                          top_paths=5)
```
tf versions: tested on both 2.4 and 2.5

However here is a sample decoded array:

(as a reminder, the alphabet is equal to [**"abcdefghijklmnopqrstuvwxyz' @"**](https://github.com/ARM-software/ML-zoo/blob/master/models/speech_recognition/wav2letter/tflite_int8/scripts/create_labels.py))

```
 print("".join([alphabet[ch_ind] for ch_ind in np.array(decoded[0].values)])) :


 oe aie bm'abaqabaiebajakyiepaubacacua ara 

```
P.S. the provided model is correct as the model prediction for the provided input matches to the provided output. 

====================================================================================
And here are my questions:

1) is the decoding process correct from your side?
2) what's your idea about decoded output?
3) is the given input a fake generated tensor?
4) could you please share your validated decoding process?
5) could you please share your feature extraction process and its corresponding preprocessing steps (i.e., normalization, etc.), in case that a raw audio wave is being used for inference.

Thanks in advance,

**Comment by tom-arm** [comment]
Hi @Ir0098,

To answer your questions:
1. Based on the code you have provided, the dequantization code seems correct.
2. The output is random as the testing input is also random.
3. Yes.

To help answer 4 & 5 for you, here is guide which demonstrates how you use the Wav2letter model end-to-end: https://developer.arm.com/documentation/102470/latest/. I hope this helps!

Thanks,
Tom

**Comment by tom-arm** [comment]
Closing this issue as it appears to be answered. Please re-open this issue if this not the case.
